superset-frontend/src/dashboard/components/gridComponents/ChartHolder.tsx

```diff
@@ -0,0 +1,333 @@
+import { ResizeCallback, ResizeStartCallback } from 're-resizable';
+import cx from 'classnames';
+import { useSelector } from 'react-redux';
+
+import { LayoutItem, RootState } from 'src/dashboard/types';
+import AnchorLink from 'src/dashboard/components/AnchorLink';
+import Chart from 'src/dashboard/containers/Chart';
+import DeleteComponentButton from 'src/dashboard/components/DeleteComponentButton';
+import DragDroppable from 'src/dashboard/components/dnd/DragDroppable';
+import HoverMenu from 'src/dashboard/components/menu/HoverMenu';
+import ResizableContainer from 'src/dashboard/components/resizable/ResizableContainer';
+import getChartAndLabelComponentIdFromPath from 'src/dashboard/util/getChartAndLabelComponentIdFromPath';
+import useFilterFocusHighlightStyles from 'src/dashboard/util/useFilterFocusHighlightStyles';
+import { COLUMN_TYPE, ROW_TYPE } from 'src/dashboard/util/componentTypes';
+import {
+  GRID_BASE_UNIT,
+  GRID_GUTTER_SIZE,
+  GRID_MIN_COLUMN_COUNT,
+  GRID_MIN_ROW_UNITS,
+} from 'src/dashboard/util/constants';
+
+export const CHART_MARGIN = 32;
+
+interface ChartHolderProps {
+  id: string;
+  parentId: string;
+  dashboardId: number;
+  component: LayoutItem;
+  parentComponent: LayoutItem;
+  getComponentById?: (id?: string) => LayoutItem | undefined;
+  index: number;
+  depth: number;
+  editMode: boolean;
+  directPathLastUpdated?: number;
+  fullSizeChartId: number | null;
+  isComponentVisible: boolean;
+
+  // grid related
+  availableColumnCount: number;
+  columnWidth: number;
+  onResizeStart: ResizeStartCallback;
+  onResize: ResizeCallback;
+  onResizeStop: ResizeCallback;
+
+  // dnd
+  deleteComponent: (id: string, parentId: string) => void;
+  updateComponents: Function;
+  handleComponentDrop: (...args: unknown[]) => unknown;
+  setFullSizeChartId: (chartId: number | null) => void;
+  postAddSliceFromDashboard?: () => void;
+}
+
+const ChartHolder: React.FC<ChartHolderProps> = ({
+  id,
+  parentId,
+  component,
+  parentComponent,
+  index,
+  depth,
+  availableColumnCount,
+  columnWidth,
+  onResizeStart,
+  onResize,
+  onResizeStop,
+  editMode,
+  isComponentVisible,
+  dashboardId,
+  fullSizeChartId,
+  getComponentById = () => undefined,
+  deleteComponent,
+  updateComponents,
+  handleComponentDrop,
+  setFullSizeChartId,
+  postAddSliceFromDashboard,
+}) => {
+  const { chartId } = component.meta;
+  const isFullSize = fullSizeChartId === chartId;
+
+  const focusHighlightStyles = useFilterFocusHighlightStyles(chartId);
+  const dashboardState = useSelector(
+    (state: RootState) => state.dashboardState,
+  );
+  const [extraControls, setExtraControls] = useState<Record<string, unknown>>(
+    {},
+  );
+  const [outlinedComponentId, setOutlinedComponentId] = useState<string>();
+  const [outlinedColumnName, setOutlinedColumnName] = useState<string>();
+  const [currentDirectPathLastUpdated, setCurrentDirectPathLastUpdated] =
+    useState(0);
+
+  const directPathToChild = useMemo(
```

Comment:
I might be mistaken, but since we are not computing an expensive function, just accessing a property inside the `dashboardState` object, could we get rid of `useMemo`? Or there is some performance implication if we remove it, that I'm not seeing. Same with `directPathLastUpdated` below.

Reply:
These two are used in other hooks/callbacks below, so need to memoize those to avoid recomputing everything (even if it's just a string/etc)